02-peer-protocol.md

```diff
@@ -49,11 +68,468 @@ transaction is confirmed are also not persistent - until you know the script
 pubkey corresponding to the funding output nothing prevents duplicative channel
 ids.
 
+### `channel_id`, v2
 
-## Channel Establishment
+For channels established using the v2 protocol, the `channel_id` is the
+SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),
+where the lesser and greater is based off the order of the basepoint.
+
+If the peer's revocation basepoint is unknown (e.g. `open_channel2`
+and `accept_channel2`), a temporary `channel_id` should be found by
+using a zeroed out basepoint for the unknown peer.
+
+#### Rationale
+
+These values must be remembered by both peers for correct operation anyway.
+They're known from the first exchange of messages, obviating the need for
+a `temporary_channel_id`. Finally, by mixing information from both sides,
+they avoid `channel_id` collisions.
+
+## Interactive Transaction Construction
+
+Interactive transaction construction allows two peers to collaboratively
+build a transaction for broadcast.  This protocol is the foundation
+for dual-funded channels establishment (v2).
+
+### Set-Up and Vocabulary
+
+There are two parties to a transaction construction: an *initiator*
+and a *non-initiator*.
+The *initiator* is the peer which initiates the protocol, e.g.
+for channel establishment v2 the *initiator* would be the peer which
+sends `open_channel2`.
+
+The protocol makes the following assumptions:
+
+- The `feerate` for the transaction is known.
+- The `dust_limit` for the transaction is known.
+- The `nLocktime` for the transaction is known.
+- The transaction version is 2.
+
+### Fee Responsibility
+
+The *initiator* is responsible for paying the fees for the following fields,
+to be referred to as the `common fields`.
+
+  - version
+  - segwit marker + flag
+  - input count
+  - output count
+  - locktime
+
+The rest of the transaction bytes' fees are the responsibility of
+the peer who contributed that input or output via `tx_add_input` or
+`tx_add_output`, at the agreed upon `feerate`.
+
+### Overview
+
+The *initiator* initiates the interactive transaction construction
+protocol with `tx_add_input`. The *non-initiator* responds with any
+of `tx_add_input`, `tx_add_output`, `tx_remove_input`, `tx_remove_output`, or
+`tx_complete`. The protocol continues with the synchronous exchange
+of interactive transaction protocol messages until both nodes have sent
+and received a consecutive `tx_complete`. This is a turn-based protocol.
+
+Once peers have exchanged consecutive `tx_complete`s, the
+interactive transaction construction protocol is considered concluded.
+Both peers should construct the transaction and fail the negotiation
+if an error is discovered.
+
+This protocol is expressly designed to allow for parallel, multi-party
+sessions to collectively construct a single transaction. This preserves
```

Comment:
wouldn't the strict ordering requirements of who sends their signatures first cause deadlocks?

i imagine 3 nodes (A, B and C) setting up a triangle of channels where every node contributes the same amount of sats.
when tx_sig has to be sent:
one node (A) has the lexicographically lowest public key and must start, now node B expects tx_sig from A, which include signatures for C's inputs, but A cant send it because it doesnt have them. it also wont get them from C because C waits until A sends tx_sig which contains signatures for B's inputs, and so on...

Reply:
Some notes on the subject written down here: https://github.com/lightning/bolts/pull/863/files#r897441047

This PR is meant to add strict ordering of tx_sig: https://github.com/rustyrussell/lightning-rfc/pull/9

It's applicable to splicing and dual not 100% clear where it should go. Putting it all in dual may be the right way to go.

Reply:
This has implications for `tx_abort` as well.

Reply:
This is a good point @antonilol, in multi-party scenarios the ordering requirements can lead to a deadlock.
I think that the simplest way to work around that is to have:

- nodes support receiving remote signatures early (eclair already does that): even if you're supposed to send first, if your remote sends their signatures first, you should gladly accept them
- when a node tries to combine multiple `interactive-tx` sessions, they should send their sigs first, which removes that deadlock

Does that make sense? It's simple enough and preserves the strict ordering for 2-party txs, which is going to be the majority of cases.

Reply:
> * nodes support receiving remote signatures early (eclair already does that): even if you're supposed to send first, if your remote sends their signatures first, you should gladly accept them.

i think the following must be changed then (and rename "strict ordering"):
```
The receiving node [of `tx_signatures`]:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by
      `tx_add_input` values
```


> * when a node tries to combine multiple `interactive-tx` sessions, they should send **their sigs** first, which removes that deadlock

**their sigs** includes sigs from the other parties.
in a case with A, B and C where A combines interactive tx sessions with B and C, B expected C's sigs from A, (and doesn't know they are from C), and the other way around.

why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers

Reply:
> in a case with A, B and C where A combines interactive tx sessions with B and C, B expected C's sigs from A, (and doesn't know they are from C), and the other way around.

Right, good catch, I was only focusing on the case where A pipes from B to C, but not the other way around as well (so my scenario actually didn't make any sense).

> why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers

Meh, this complicates the state machine for no good reason...better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can, don't you think?

Splitting the sigs on a per-input basis wouldn't fundamentally fix the issue. If A is supposed to start in her negotiation with B, if she sends signatures for her inputs but not the inputs she imported from C, why would B start sending their sigs? If there's an ordering requirement, B would wait to receive **all** sigs before sending any of theirs....

I think we don't have a choice, having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement. @niftynei @ddustin WDYT?

Reply:
> better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can

this does not fix it completely, this would work in the case where there are no loops in nodes having a interactive tx session. if nodes have to send all signatures at once at there happens to be a loop they still wait on each other.
when nodes A, B and C batch open a few channels at the same time (and use interactive tx with others to minimize linkability on chain and save on fees) without them knowing of each other, they all wait on the other parties.

i think both are needed:
* no strict ordering
* ability to send signatures separately

Reply:
Can you detail why we'd want to allow loops?
Can that even happen?
It feels to me that it cannot happen because you would end up with a transaction spending the same input twice?
Please detail an example flow to make sure we're on the same page.

Reply:
Can you detail why we'd want to allow loops?
I'm unsure whether they're desirable, and I'm unsure whether they create a deadlock or not.
Please detail an example flow to make sure we're on the same page.

Reply:
> > why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers
> 
> Meh, this complicates the state machine for no good reason...better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can, don't you think?

The spec as it is has you send all signatures you're responsible at once in a variable size tlv https://github.com/lightning/bolts/pull/851/files#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR360-R370 (line 360-370 in 02-peer-protocol.md).

I like this approach and have found it to work well for me. This avoids extra rounds of `tx_signature` back-and-forth that add protocol complications. I can't see any benefit of switching it out for rounds 🤷.

> I think we don't have a choice, having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement. @niftynei @ddustin WDYT?

The conclusion @niftynei and I came to is essentially this:

`tx_signature` griefing is most useful to the side that is sending less sats. By requiring the side with the least sats being added with `tx_add_input` to sign first we eliminate this griefing potential.

```
        +-------+                              +-------+                               +-------+
        |       |-- add_input 200sat --------->|       |<-- add_input 100sat ----------|       |
        |       |<- add_input 10sat -----------|       |--- add_input 10sat ---------->|       |
        |       |<- relay_C_add_input 100sat --|       |-- relay_A_add_input 200sat -->|       |
        |       |                              |       |                               |       |
        |   A   |                              |   B   |                               |   C   |
        |       |                              |       |<-- tx_signature --------------|       |
        |       |<-- tx_signature -------------|       |                               |       |
        |       |--- tx_signature ------------>|       |                               |       |
        |       |                              |       |--- tx_signature ------------->|       |
        +-------+                              +-------+                               +-------+
```

An important note here: when calculating how many sats are on your side you include any inputs you're relaying for others. From the perspective of your peer there is no difference between your own inputs and relayed inputs.

In this scenario C signs first because it sees itself adding 100 sats vs 210 sats coming in.

After receiving C's signatures, B signs next because 110 sats is less than 200 sats coming from A. It forwards this signature plus C's signatures to A in a single `tx_signature` message.

Then A signs and sends `tx_signature` along to B.

Finally, B bundles A's signatures with its own and sends them to C.

> having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement.

Not having an ordering requirement is what actually introduces the deadlock potential -- the incentive is to always sign last. Having a strict ordering of who must sign first that is whomever has the least to lose is designed to mitigate it.

Reply:
> i imagine 3 nodes (A, B and C) setting up a triangle of channels where every node contributes the same amount of sats.

fwiw this is impossible, as it'd break the rule that you fail the interactive tx negotiation if a node sends you an input that's yours. (as C would ostensibly do to A).

You could make 2 channels (A-B-C) but not a circle (A-B-C-A).

Reply:
I thought about this some more, and I agree with @niftynei and @ddustin. @antonilol feel free to share an example scenario in case I misunderstood your point, but I don't think what you wanted to achieve would work, regardless of signatures ordering. Loops are simply not possible, as only one node at a time can batch multiple channel opens into a single tx, otherwise there would be a deadlock waiting for `tx_complete` (so even before we reach the signatures step). 

Here is why:

- the full transaction must be the same in all parallel sessions (otherwise signatures can't be transferred from one session to another)
- so a node that wants to batch multiple sessions in a single tx needs to wait for their peers to send `tx_complete` before sending their own, because they need to have the full set of inputs/outputs of each peer to transfer them to the other sessions
- this batching node will probably "cheat" by adding and removing a dummy input until they receive `tx_complete` from their peers
- if both nodes in a session do that, it will never converge and nobody will send `tx_complete`
- when that happens, negotiation will fail after 4096 iterations of adding an input or output

Also, the ordering requirement based on amount works well for this kind of batching. Let's consider B batching a session with A and a session with C. From A's point of view, B is adding `amount_BA + amount_CB` to the channel. From C's point of view, B is adding `amount_BC + amount_AB` to the channel. For a deadlock at the signing step to occur, we would need to have both `amount_BA + amount_CB <= amount_AB` and `amount_BC + amount_AB <= amount_CB`, which implies `amount_BA + amount_CB <= amount_CB - amount_BC` => `amount_BA <= -amount_BC` which is only possible if B contributes nothing to the channels...so as long as the coordinator contributes a non-zero amount to at least one channel, there is a channel where they won't start sending their sigs, so they won't be blocked. Let me know if you want me to formalize it better (I believe this generalizes to N sessions but haven't written it down yet, will do today).

Maybe the spec should mention that only the initiator should try to batch multiple sessions, otherwise there's a risk of deadlock?

Reply:
> I believe this generalizes to N sessions but haven't written it down yet, will do today

It does generalize to N sessions, as long as the coordinator contributes a non-zero amount to each channel, there can't be a deadlock. I can share the (small) proof if you're interested.

Reply:
fwiw I believe @ddustin and I independently reached this same conclusion wrt "no deadlock as long as the coordinator contributes funds."

Reply:
> @antonilol feel free to share an example scenario in case I misunderstood your point, but I don't think what you wanted to achieve would work, regardless of signatures ordering. Loops are simply not possible, as only one node at a time can batch multiple channel opens into a single tx, otherwise there would be a deadlock waiting for `tx_complete` (so even before we reach the signatures step).

when making it i came to the same `tx_complete` deadlock. i would work though if all collaborating parties know that they all work on the same tx and have some other medium to send messages (to coordinate who sets it up and to not make loops)

how should accidental loops be handled? multiple nodes that work on the same tx (but not all) will notice it when their in/output gets sent back and fail the negotiation. others will notice just that their peer disconnected after they sent a in/output (that was theirs).

Reply:
> how should accidental loops be handled?

If only the initiator does that kind of batching, there is no risk of accidental loops. I would simply recommend against trying to batch when you're not the initiator, you'll always run the risk of creating those loops otherwise.

> how should accidental loops be handled?

When a node detects a loop, they will send `tx_abort` (with a custom message that may explain why they're aborting) to cancel everything.

If there is a deadlock before `tx_complete`, the protocol will abort after 4096 iterations as well, and this is a good indication that both sides of an interactive-tx session are trying to batch at the same time, which cannot work.